sdk/iot/hub/samples/twin/src/iot_hub_twin_example.c

```diff
@@ -0,0 +1,447 @@
+  mqtt_ssl_options.keyStore = (char*)x509_cert_pem_file;
+  if (*x509_trust_pem_file != '\0')
+  {
+    mqtt_ssl_options.trustStore = (char*)x509_trust_pem_file;
+  }
+
+  mqtt_connect_options.ssl = &mqtt_ssl_options;
+
+  if ((rc = MQTTClient_connect(mqtt_client, &mqtt_connect_options)) != MQTTCLIENT_SUCCESS)
+  {
+    printf("Failed to connect, return code %d\n", rc);
+    return rc;
+  }
+
+  return 0;
+}
+
+static int subscribe()
+{
+  int rc;
+
+  char twin_desired_topic[128];
+  if ((rc = az_iot_hub_client_twin_patch_get_subscribe_topic_filter(
+           &client, twin_desired_topic, sizeof(twin_desired_topic), NULL))
+      != AZ_OK)
+
+  {
+    printf("Failed to get C2D MQTT SUB topic filter, return code %d\n", rc);
+    return rc;
+  }
+
+  char twin_response_topic[128];
+  if ((rc = az_iot_hub_client_twin_response_get_subscribe_topic_filter(
+           &client, twin_response_topic, sizeof(twin_response_topic), NULL))
+      != AZ_OK)
+
+  {
+    printf("Failed to get C2D MQTT SUB topic filter, return code %d\n", rc);
+    return rc;
+  }
+
+  if ((rc = MQTTClient_subscribe(mqtt_client, twin_desired_topic, 1)) != MQTTCLIENT_SUCCESS)
+  {
+    printf("Failed to subscribe, return code %d\n", rc);
+    return rc;
+  }
+  if ((rc = MQTTClient_subscribe(mqtt_client, twin_response_topic, 1)) != MQTTCLIENT_SUCCESS)
+  {
+    printf("Failed to subscribe, return code %d\n", rc);
+    return rc;
+  }
+
+  return 0;
+}
+
+static int send_get_twin()
+{
+  int rc;
+  printf("Requesting twin document\n");
+
+  //Get the topic if we haven't already
+  if (get_twin_topic[0] == '\0')
+  {
+    if ((rc = az_iot_hub_client_twin_document_get_publish_topic(
+             &client,
+             get_twin_topic_request_id,
+             get_twin_topic,
+             sizeof(get_twin_topic),
+             NULL))
+        != AZ_OK)
+    {
+      printf("Unable to get twin document publish topic, return code %d\n", rc);
+      return rc;
+    }
+  }
+
+  if ((rc = MQTTClient_publish(mqtt_client, get_twin_topic, 0, NULL, 0, 0, NULL))
+      != MQTTCLIENT_SUCCESS)
+  {
+    printf("Failed to publish twin document request, return code %d\n", rc);
+    return rc;
+  }
+  return rc;
+}
+
+static int build_reported_property(az_json_builder* json_builder)
+{
+  az_result result;
+  result = az_json_builder_init(json_builder, AZ_SPAN_FROM_BUFFER(reported_property_payload));
+  result = az_json_builder_append_token(json_builder, az_json_token_object_start());
+  az_json_token reported_property_value_token
+      = az_json_token_number((double)reported_property_value++);
+  result = az_json_builder_append_object(
+      json_builder, reported_property_name, reported_property_value_token);
+  result = az_json_builder_append_token(json_builder, az_json_token_object_end());
+
+  return result;
+}
+
+static int send_reported_property()
+{
+  int rc;
+  printf("Sending reported property\n");
+
+  //Get the topic if we haven't already
+  if (reported_property_topic[0] == '\0')
```

Comment:
Same as above: we should remove this check. Also, the topic should always be constructed in a way that ensures that the request_id is different.

Reply:
ah true. was trying to optimize. will change